jaeger-jaxrs2/src/test/java/com/uber/jaeger/propagation/ServerFilterIntegrationTest.java

```diff
@@ -0,0 +1,197 @@
+
+  private TraceContext clientTraceContext = new ActiveSpanSourceTraceContext(clientTracer);
+
+  private HttpServer server;
+
+  @Override
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+
+    // start the server
+    final ResourceConfig rc = (ResourceConfig) configure();
+    server = GrizzlyHttpServerFactory.createHttpServer(URI.create(BASE_URI), rc);
+  }
+
+  @Override
+  @After
+  public void tearDown() throws Exception {
+    super.tearDown();
+
+    // stop the server
+    if (server.isStarted()) {
+      server.stop();
+    }
+  }
+
+  @Override
+  protected Application configure() {
+
+    io.opentracing.Tracer serverTracer =
+        new com.uber.jaeger.Tracer
+            .Builder("server-tracer", new NoopReporter(), new ConstSampler(true))
+            .build();
+
+    TraceContext serverTraceContext = new ActiveSpanSourceTraceContext(serverTracer);
+    
+    // Our 'Application'
+    ResourceConfig resourceConfig =
+        new ResourceConfig()
+            .register(new ServerFilter(serverTracer, serverTraceContext), Priorities.USER + 10)
+            .register(new TestHookFilter(serverTraceContext), Priorities.USER + 20)
+            .register(TestResource.class);
+
+    return resourceConfig;
+  }
+
+  @After
+  public void teardownTestingInfra() {
+    // stop server
+    if (server.isStarted()) {
+      server.stop();
+    }
+  }
+
+  @Test
+  public void testTracingUsingInstrumentedClient() {
+    com.uber.jaeger.Span span = (com.uber.jaeger.Span) clientTracer
+        .buildSpan("root-span")
+        .startManual();
+    span.setBaggageItem("x-test", "testing");
+    clientTraceContext.push(span);
+
+    Response resp = target()
+        .register(new ClientFilter(clientTracer, clientTraceContext))
+        .path("test")
+        .request()
+        .header(DEBUG_JAEGER_SPAN_BAGGAGE_HEADER, "foo=bar")
+        .get();
+    assertEquals(resp.getStatus(), 200);
+  }
+
+  @Test
+  public void testTracingUsingInstrumentedClientWithFakeBaggageHeader() {
+    Response resp = target()
+        .register(new ClientFilter(clientTracer, clientTraceContext))
+        .path("test")
+        .request()
+        .header(DEBUG_JAEGER_SPAN_BAGGAGE_HEADER, "foo=bar")
+        .header("uberctx-x-test", "testing")
+        .get();
+    assertEquals(resp.getStatus(), 200);
+  }
+
+  @Test
+  public void testTracingUsingNonInstrumentedClient() {
+    Response resp = target()
+        .path("test")
+        .request()
+        .header(DEBUG_ID_HEADER_KEY, "12345")
+        .header(DEBUG_JAEGER_SPAN_BAGGAGE_HEADER, "foo=bar,x-test=testing")
+        .get();
+    assertEquals(resp.getStatus(), 200);
+  }
+
+  @Test
+  public void testTracingUsingNonInstrumentedClientWithFakeBaggageHeader() {
+    Response resp = target()
+        .path("test")
+        .request()
+        .header(DEBUG_ID_HEADER_KEY, "12345")
+        .header(DEBUG_JAEGER_SPAN_BAGGAGE_HEADER, "foo=bar")
+        .header("uberctx-x-test", "testing")
```

Comment:
one convenient way to write these kinds of tests is to make the server simply return the values in a JSON, and assert them in the tests. For example, in this test it's not clear if the debug-id was actually processed, i.e. that the span debug flag was set and the header value was stored as a tag.